File: 백준/Gold/2096. 내려가기/내려가기.java
import java.io.*;

public class Main {

    static StreamTokenizer st = new StreamTokenizer(new InputStreamReader(System.in));

    static int readInt() throws IOException {
        st.nextToken();
        return (int) st.nval;
    }

    public static void main(String[] args) throws IOException {
        int n = readInt();
        int[] minCum = new int[3];
        int[] maxCum = new int[3];
        int[] minWin = {readInt(), readInt(), readInt()};
        int[] maxWin = minWin.clone();

        for (int i = 1; i < n; i++) {
            for (int c = 0; c < 3; c++) {
                minCum[c] = minWin[c];
                maxCum[c] = maxWin[c];

                minWin[c] = readInt();
                maxWin[c] = minWin[c];
            }

            for (int c = 0; c < 3; c++) {
                minWin[c] += minWith(minCum, c);
                maxWin[c] += maxWith(maxCum, c);
            }
        }

        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        bw.append(String.format("%d %d", maxWith(maxWin, 1), minWith(minWin, 1)));
        bw.flush();
        bw.close();
    }

    static int minWith(int[] arr, int c) {
        if (c == 1) {
            return Math.min(arr[0], Math.min(arr[1], arr[2]));
        }

        return Math.min(arr[c], arr[1]);
    }

    static int maxWith(int[] arr, int c) {
        if (c == 1) {
            return Math.max(arr[0], Math.max(arr[1], arr[2]));
        }

        return Math.max(arr[c], arr[1]);
    }
}

File: 백준/Gold/2096. 내려가기/내려가기.kt
import java.io.StreamTokenizer

fun main() = with(StreamTokenizer(System.`in`.bufferedReader())) {
    fun readInt(): Int {
        nextToken()
        return nval.toInt()
    }

    val n = readInt()
    val winMins = IntArray(3) { readInt() }
    val winMaxs = winMins.copyOf()
    val mins = winMins.copyOf()
    val maxs = winMins.copyOf()

    repeat(n - 1) {
        repeat(3) { i ->
            winMins[i] = mins[i]
            winMaxs[i] = maxs[i]

            mins[i] = readInt()
            maxs[i] = mins[i]
        }

        for (c in 0..2) {
            mins[c] += when (c) {
                0 -> minOf(winMins[0], winMins[1])
                2 -> minOf(winMins[1], winMins[2])
                else -> winMins.min()
            }

            maxs[c] += when (c) {
                0 -> maxOf(winMaxs[0], winMaxs[1])
                2 -> maxOf(winMaxs[1], winMaxs[2])
                else -> winMaxs.max()
            }
        }
    }

    with(System.out.bufferedWriter()) {
        append("${maxs.max()} ${mins.min()}")
        flush()
        close()
    }
}